src/inc/RingBuffer.h

```diff
@@ -0,0 +1,381 @@
+    {
+        int lowerBound = indexWrapper(static_cast<int>(fOffset));
+        int upperBound = indexWrapper(lowerBound + 1);
+
+        // determine if the upper bound should be enlarged.
+        float fractionPart = abs(fOffset - static_cast<int>(fOffset));
+        
+        if (!(0 <= fractionPart && fractionPart <= 1)) {
+            std::printf("%.8f %.8f\n", fOffset, fractionPart);
+        }
+
+        if (upperBound >= m_iBuffLength) {
+            throw std::exception("The input offset is too big and exceeds the buffer length.");
+        }
+
+        // interpolate between lower & upper bound
+        T fLowerBoundValue = get(lowerBound);
+        T fUpperBoundValue = get(upperBound);
+
+        return (1 - fractionPart) * fLowerBoundValue + fractionPart * fUpperBoundValue;
+    }
+
+    /*! set buffer content and indices to 0
+    \return void
+    */
+    void reset()
+    {
+        std::memset (m_ptBuff, 0, sizeof(T)*m_iBuffLength);
+        m_iReadIdx  = 0;
+        m_iWriteIdx = 0;
+    }
+
+    /*! return the current index for writing/put
+    \return int
+    */
+    int getWriteIdx() const
+    {
+        return m_iWriteIdx;
+    }
+
+    /*! move the write index to a new position
+    \param iNewWriteIdx: new position
+    \return void
+    */
+    void setWriteIdx(int iNewWriteIdx)
+    {
+        incIdx(m_iWriteIdx, iNewWriteIdx - m_iWriteIdx);
+    }
+
+    /*! return the current index for reading/get
+    \return int
+    */
+    int getReadIdx() const
+    {
+        return m_iReadIdx;
+    }
+
+    /*! move the read index to a new position
+    \param iNewReadIdx: new position
+    \return void
+    */
+    void setReadIdx(int iNewReadIdx)
+    {
+        incIdx(m_iReadIdx, iNewReadIdx - m_iReadIdx);
+    }
+
+    /*! returns the number of values currently buffered (note: 0 could also mean the buffer is full!)
+    \return int
+    */
+    int getNumValuesInBuffer() const
+    {
+        return (m_iWriteIdx - m_iReadIdx + m_iBuffLength) % m_iBuffLength;
+    }
+
+    /*! returns the length of the internal buffer
+    \return int
+    */
+    int getLength() const
+    {
+        return m_iBuffLength;
+    }
+
+    int indexWrapper(int index) const {
+        return (index < 0) ? m_iBuffLength - (-index % m_iBuffLength) : index % m_iBuffLength;
+    }
+
+private:
+    CRingBuffer();
+    CRingBuffer(const CRingBuffer& that);
+
+    void incIdx(int& iIdx, int iOffset = 1)
+    {
+        while ((iIdx + iOffset) < 0)
+        {
+            // avoid negative buffer indices
+            iOffset += m_iBuffLength;
+        }
+        iIdx = (iIdx + iOffset) % m_iBuffLength;
+    };
+
+    int m_iBuffLength = 0,      //!< length of the internal buffer
+        m_iReadIdx = 0,         //!< current read index
+        m_iWriteIdx = 0;        //!< current write index
+
+    T* m_ptBuff = 0;            //!< data buffer
+};
+#endif // __RingBuffer_hdr__
+
+
+// #if !defined(__RingBuffer_hdr__)
```

Comment:
The rest can be deleted

Reply:
I will delete the existing version and substitute with the currently commented part. Still you are correct, I should keep code base clean.